Initialize upgrader and operator version files

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69427
* #68773
* __->__ #68772

Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

tugsbayasgalan committed:
Initialize upgrader and operator version files

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603257).

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

gmagogsfm commented on torch/jit/operator_upgraders.py:
```diff
@@ -0,0 +1,99 @@
+    return formatted_table
+
+def generate_bytecode(file_name):
+    yaml_content = [
+        {"div_Tensor_0_3": format_bytecode(torch._C._compile_graph_to_code_table("div_Tensor_0_3", div_Tensor_0_3.graph))},
+        {"div_Scalar_0_3": format_bytecode(torch._C._compile_graph_to_code_table("div_Scalar_0_3", div_Scalar_0_3.graph))},
+        {"div_out_0_3": format_bytecode(torch._C._compile_graph_to_code_table("div_out_0_3", div_out_0_3.graph))},
+        {"div__Tensor_0_3": format_bytecode(torch._C._compile_graph_to_code_table("div__Tensor_0_3", div__Tensor_0_3.graph))},
+        {"div__Scalar_0_3": format_bytecode(torch._C._compile_graph_to_code_table("div__Scalar_0_3", div__Scalar_0_3.graph))},
```
Why do we need to enumerate yaml_content when we have the file `operator_versions.yaml` ready, which has identical info?

tugsbayasgalan replied:
Sorry operator_versions.yaml will be removed. One thing I can do is I can access `operator_version_map` from version_map.h through Pybind and generate the bytecode from that map instead of enumerating there. By doing so, we can also ensure there is 1 to 1 mapping between `operator_version_map` and the list of entries here.

gmagogsfm replied:
That should work.

Could you also give me a list of places that a developer need to touch when adding a new upgrader? Ideally there should be just one place

tugsbayasgalan replied:
So far torch/jit/operator_upgraders.py and torch/csrc/jit/operator_upgraders/version_map.h

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

tugsbayasgalan commented on torch/jit/operator_upgraders.py:
```diff
@@ -0,0 +1,99 @@
+
+# @torch.jit.script
+# def full_out_0_4(size: List[int], fill_value: Union[int, float, complex], *, out: torch.Tensor) -> torch.Tensor:
+#     return torch.full(size, fill_value, out=out)
```
this is resolved now

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603257).

tugsbayasgalan commented on torch/jit/operator_upgraders.py:
```diff
@@ -0,0 +1,99 @@
+1. The body of the function must be Torchscriptable
+2. The naming convention of the upgraders should be:
+    <op_name>_<op_overload>_upgrader_<old_version>_<new_version>
+3. The name of the upgrader must be present in operator_versions.yaml
```
because we are relying on the bumped version number, I don't think we really don't need to strictly enforce this naming convention.

tugsbayasgalan commented on torch/jit/operator_upgraders.py:
```diff
@@ -0,0 +1,102 @@
+        return self.true_divide(other)
+    return self.divide(other, rounding_mode='trunc')
+
+# # TODO: not present in the schema
```
I am assuming its safe to not have upgraders here. The current version of this function does exact same thing as this one here. (@gmagogsfm)

gmagogsfm replied:
"The current version of this function does exact same thing as this one here."

What are "this function" and "this one"?

tugsbayasgalan replied:
This function (aka operator) -> the upgrader for `"def div_0_3(self: number, other: number) -> number"`

this one -> (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/register_prim_ops.cpp#L806) (the current version of this operator)

gmagogsfm replied:
Yeah, they do seem identical. `/` is compiled to `aten::div` anyway, so this is a no op. Wonder why this was added in the first place.

tugsbayasgalan replied:
Resolved offline

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603257).

cccclai commented on torch/csrc/jit/operator_upgraders/version_map.h:
```diff
@@ -0,0 +1,51 @@
+};
+
+static std::unordered_map<std::string, std::vector<UpgraderEntry>> operator_version_map(
+    {{"div.Tensor",
```
How do you swap the operator in the graph? When I try to replace the operator in bytecode, the operator name will be `aten::div.Tensor`, instead of `div.Tensor`.

tugsbayasgalan replied:
good catch! It should be aten::div.Tensor

cccclai replied:
Oh ok. Thanks. I manually add it in the codegen file in https://github.com/pytorch/pytorch/pull/69194. Will remove the manual part after this pr is updated.

Commit:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/operator_upgraders/version_map.h:
```diff
@@ -0,0 +1,51 @@
+  std::string old_schema;
+};
+
+static std::unordered_map<std::string, std::vector<UpgraderEntry>> operator_version_map(
```
```suggestion
const static std::unordered_map<std::string, std::vector<UpgraderEntry>> kOperatorVersionMap(
```
Consider using const since it won't change. Then naming convention will be `kOperatorVersionMap`.

tugsbayasgalan replied:
What does k in front of the variable stand for?

cccclai replied:
const

tugsbayasgalan replied:
i see thx

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603257).

tugsbayasgalan committed:
Update on "Initialize upgrader and operator version files"


Differential Revision: [D32603257](https://our.internmc.facebook.com/intern/diff/D32603257)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603257).
